File: utilities/c/get_input.c
#include "get_input.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INITIAL_CAPACITY 100

InputLines* get_input(int day) {
    char* path = get_input_path(day);
    InputLines* result = read_input(path);
    free(path);
    return result;
}

InputLines* get_test_input(int day, int test_num) {
    char* path = get_test_input_path(day, test_num);
    InputLines* result = read_input(path);
    free(path);
    return result;
}

char* get_input_path(int day) {
    // Solutions are in days/day_NN/solutions/
    // Input is in days/day_NN/data/input.txt
    char* path = (char*)malloc(256 * sizeof(char));
    snprintf(path, 256, "../data/input.txt");
    return path;
}

char* get_test_input_path(int day, int test_num) {
    char* path = (char*)malloc(256 * sizeof(char));
    snprintf(path, 256, "../data/test_%d.txt", test_num);
    return path;
}

InputLines* read_input(const char* file_path) {
    FILE* file = fopen(file_path, "r");
    if (file == NULL) {
        return NULL;
    }

    InputLines* input = (InputLines*)malloc(sizeof(InputLines));
    input->count = 0;
    input->capacity = INITIAL_CAPACITY;
    input->lines = (char**)malloc(input->capacity * sizeof(char*));

    char line[MAX_LINE_LENGTH];
    while (fgets(line, sizeof(line), file)) {
        // Trim newline
        size_t len = strlen(line);
        if (len > 0 && line[len - 1] == '\n') {
            line[len - 1] = '\0';
        }

        // Expand capacity if needed
        if (input->count >= input->capacity) {
            input->capacity *= 2;
            input->lines = (char**)realloc(input->lines, input->capacity * sizeof(char*));
        }

        // Allocate and copy line
        input->lines[input->count] = (char*)malloc((strlen(line) + 1) * sizeof(char));
        strcpy(input->lines[input->count], line);
        input->count++;
    }

    fclose(file);
    return input;
}

char* read_input_raw(const char* file_path) {
    FILE* file = fopen(file_path, "r");
    if (file == NULL) {
        return NULL;
    }

    fseek(file, 0, SEEK_END);
    long size = ftell(file);
    fseek(file, 0, SEEK_SET);

    char* content = (char*)malloc((size + 1) * sizeof(char));
    fread(content, 1, size, file);
    content[size] = '\0';

    fclose(file);
    return content;
}

void free_input_lines(InputLines* input) {
    if (input == NULL) {
        return;
    }

    for (int i = 0; i < input->count; i++) {
        free(input->lines[i]);
    }
    free(input->lines);
    free(input);
}

void free_input_raw(char* content) {
    if (content != NULL) {
        free(content);
    }
}

File: utilities/go/get_input.go
package utils

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// GetInput reads input from input.txt for the given day.
// day should be a number like 1, 2, 3, etc.
// Returns lines as a slice of strings.
func GetInput(day int) ([]string, error) {
	path := GetInputPath(day)
	return ReadInput(path)
}

// GetTestInput reads input from test_N.txt for the given day and test number.
// day should be a number like 1, 2, 3, etc.
// testNum should be the test number (1, 2, 3, etc.)
// Returns lines as a slice of strings.
func GetTestInput(day int, testNum int) ([]string, error) {
	path := GetTestInputPath(day, testNum)
	return ReadInput(path)
}

// GetInputPath returns the path to input.txt for the given day.
func GetInputPath(day int) string {
	// Get the path relative to the solution file
	// Solutions are in days/day_NN/solutions/
	// Input is in days/day_NN/data/input.txt
	return fmt.Sprintf("../data/input.txt")
}

// GetTestInputPath returns the path to test_N.txt for the given day and test number.
func GetTestInputPath(day int, testNum int) string {
	return fmt.Sprintf("../data/test_%d.txt", testNum)
}

// ReadInput reads a file and returns lines as a slice of strings.
func ReadInput(filePath string) ([]string, error) {
	// Resolve relative paths from solution location
	absPath, err := filepath.Abs(filePath)
	if err != nil {
		return nil, fmt.Errorf("error resolving path: %w", err)
	}

	file, err := os.Open(absPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, strings.TrimSpace(scanner.Text()))
	}
	return lines, scanner.Err()
}

// ReadInputRaw reads a file and returns raw content as a string.
func ReadInputRaw(filePath string) (string, error) {
	absPath, err := filepath.Abs(filePath)
	if err != nil {
		return "", fmt.Errorf("error resolving path: %w", err)
	}

	data, err := os.ReadFile(absPath)
	if err != nil {
		return "", err
	}
	return string(data), nil
}
